File: src/main/java/eu/qwsome/sql/condition/ComparableFieldImpl.java
package eu.qwsome.sql.condition;

import eu.qwsome.sql.Select;

import java.util.Collection;

import static eu.qwsome.sql.SubselectValueHolder.subselect;

/**
 * @author Lukáš Kvídera
 */
public class ComparableFieldImpl implements ComparableField {

  private final ValueHolder comparedField;

  public ComparableFieldImpl(final ValueHolder comparedField) {
    this.comparedField = comparedField;
  }

  @Override
  public Condition isEqualTo(final ValueHolder another) {
    return new IsEqual(this.comparedField, another);
  }

  @Override
  public Condition isNotEqualTo(final ValueHolder another) {
    return new IsNotEqual(this.comparedField, another);
  }

  @Override
  public Condition isBetween(final ValueHolder from, final ValueHolder to) {
    return new Between(this.comparedField, from, to);
  }

  @Override
  public Condition isNull() {
    return new IsNull(this.comparedField);
  }

  @Override
  public Condition isLessThan(final ValueHolder another) {
    return new IsLessThan(this.comparedField, another);
  }

  @Override
  public Condition isLessOrEqualThan(final ValueHolder another) {
    return new IsLessOrEqualThan(this.comparedField, another);
  }

  @Override
  public Condition isNotNull() {
    return new IsNotNull(this.comparedField);
  }

  @Override
  public Condition isGreaterThan(final ValueHolder another) {
    return new IsGreaterThan(this.comparedField, another);
  }

  @Override
  public Condition isGreaterOrEqualThan(final ValueHolder another) {
    return new IsGreaterOrEqualThan(this.comparedField, another);
  }

  @Override
  public Condition in(final ValueHolder... another) {
    return new In(this.comparedField, another);
  }

  @Override
  public Condition in(final Collection<ValueHolder> another) {
    return new In(
      this.comparedField,
      another.stream().toArray(ValueHolder[]::new)
    );
  }

  @Override
  public Condition in(final Select.ConditionsBuiltPhase subselect) {
    return new In(this.comparedField, subselect(subselect));
  }

  @Override
  public Condition like(final ValueHolder pattern) {
    return new Like(this.comparedField, pattern);
  }

  @Override
  public Condition notLike(final ValueHolder pattern) {
    return new NotLike(this.comparedField, pattern);
  }

}

File: src/main/kotlin/eu/qwsome/sql/condition/ComparableFieldImpl.kt
/*
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql.condition

import eu.qwsome.sql.Select
import eu.qwsome.sql.SubselectValueHolder

class ComparableFieldImpl(private val comparedField: ValueHolder) : ComparableField {

    override fun isEqualTo(another: ValueHolder): Condition =
        IsEqual(comparedField, another)

    override fun isNotEqualTo(another: ValueHolder): Condition =
        IsNotEqual(comparedField, another)

    override fun isBetween(from: ValueHolder, to: ValueHolder): Condition =
        Between(comparedField, from, to)

    override fun isNull(): Condition = IsNull(comparedField)

    override fun isLessThan(another: ValueHolder): Condition =
        IsLessThan(comparedField, another)

    override fun isLessOrEqualThan(another: ValueHolder): Condition =
        IsLessOrEqualThan(comparedField, another)

    override fun isNotNull(): Condition =
        IsNotNull(comparedField)

    override fun isGreaterThan(another: ValueHolder): Condition =
        IsGreaterThan(comparedField, another)

    override fun isGreaterOrEqualThan(another: ValueHolder): Condition =
        IsGreaterOrEqualThan(comparedField, another)

    override fun `in`(vararg another: ValueHolder): Condition =
        In(comparedField, *another)

    override fun `in`(another: Collection<ValueHolder>): Condition =
        In(comparedField, *another.toTypedArray())

    override fun `in`(subselect: Select.ConditionsBuiltPhase): Condition =
        In(comparedField, SubselectValueHolder.subselect(subselect))

    override fun like(pattern: ValueHolder): Condition =
        Like(comparedField, pattern)

    override fun notLike(pattern: ValueHolder): Condition =
        NotLike(comparedField, pattern)
}
